frontend/src/main/scala/bloop/io/SourceWatcher.scala

```diff
@@ -61,22 +63,38 @@ final class SourceWatcher(project: Project, dirs0: Seq[Path], logger: Logger) {
       }
     )
 
-    val watchingTask = Task {
+    // Use Java's completable future because we can stop/complete it from the cancelable
+    val watcherHandle = watcher.watchAsync(ExecutionContext.ioExecutor)
+    val watchController = Task {
       logger.info(s"File watching $dirsCount directories...")
-      try watcher.watch()
+      try watcherHandle.get()
       finally watcher.close()
-    }.doOnCancel(Task {
+      logger.debug("The file watcher was successfully closed.")
+    }
+
+    val watchCancellation = Cancelable { () =>
+      watchingEnabled = false
+      watcherHandle.complete(null)
       observer.onComplete()
-      watcher.close()
       ngout.println(
         s"File watching on '${project.name}' and dependent projects has been successfully cancelled.")
-    })
+    }
 
-    val watchHandle = watchingTask.materialize.runAsync(ExecutionContext.ioScheduler)
+    val fileEventConsumer = {
+      FoldLeftAsyncConsumer.consume[State, DirectoryChangeEvent](state0) {
+        case (state, event) =>
+          event.eventType match {
+            case EventType.CREATE => runAction(state, event)
+            case EventType.MODIFY => runAction(state, event)
+            case EventType.OVERFLOW => runAction(state, event)
+            case EventType.DELETE => Task.now(state)
+          }
+      }
+    }
 
+    val watchHandle = watchController.runAsync(ExecutionContext.ioScheduler)
     observable
       .consumeWith(fileEventConsumer)
-      .doOnFinish(_ => Task(watchHandle.cancel()))
```

Comment:
@alexandru If you add `doOnFinish` here, the cancellation logic of the consumer is not executed. Is this a bug?